comments/heroku_logs_to_cloudwatch.md

```diff
@@ -0,0 +1,79 @@
+
+[Heroku's stated reason](https://devcenter.heroku.com/articles/log-drains) for providing such a small log buffer is that they expect users to "drain" the log data into a separate, dedicated log storage system that will retain the logs long-term. Heroku do not provide anything like this off-the-shelf, and although there are a number of paid third-party addons that claim to do the job we do not want to overcomplicate our Heroku deployment.
+
+Given that we already use AWS for a number of functions, if we want long-term log storage it makes sense to write a Lambda function that will ingest the logs into AWS Cloudwatch, which is Amazon's logging and metrics service.
+
+### Why does this document exist?
+
+Because while there are a few pieces of code in public Github repositories that provide a basic example of how shipping logs from Heroku to Cloudwatch via Lambda works (see [here](https://github.com/rwilcox/heroku_cloudwatch_sync/blob/master/src/heroku_sync_to_cloudwatch.py), there are others available with some basic googling), attempting to adapt this example code into an actual working Lambda function that didn't scream blue murder in our alerting channel took four rounds of refactoring. The learnings are too numerous to be captured in comments in the function code, so I'm writing them down here.
+
+## Basic operation
+
+In theory, very simple:
+
+- Write a Lambda function handler in Python capable of parsing Heroku log data and forwarding it on to Cloudwatch
+- Connect Heroku up to the Lambda function by using the Heroku CLI to add the Lambda function URL as a log drain: `heroku drains:add <function url> -a <appname>`.
+
+In practice, very difficult.
+
+### Problem 1: Cloudwatch log groups and log streams
+
+AWS Cloudwatch logging relies on the concept of log *streams*, which contain streams of individual log messages, which are all part of a single log *group*. Log streams are the things that actually get written to by the Lambda function, and the log group allows filtering and searching to be run across multiple log streams at once.
+
+Most of the example code for pushing logs to Cloudwatch only pushed the logs to a single static log stream. This unfortunately does not work in practice, because of...
+
+### Problem 2: Parallel Lambda function execution
+
+If you push five separate events up to a Lambda function, AWS will not invoke one instance of the function that processes those five events one after another. Instead it will spawn five function instances that each process one event in parallel. 
```

Comment:
It's a little non-obvious that we end up with a system with much parallelism. It would make sense for heroku to open one connection per log drain, rather than multiple, and this to generate sequential events from the lamba code's perspective. That said, this all seems fine.